helix-dap/src/client.rs

```diff
@@ -0,0 +1,352 @@
+    path_format: String,
+    supports_variable_type: bool,
+    supports_variable_paging: bool,
+    supports_run_in_terminal_request: bool,
+    supports_memory_references: bool,
+    supports_progress_reporting: bool,
+    supports_invalidated_event: bool,
+}
+
+// TODO: split out
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct LaunchArguments {
+    mode: String,
+    program: String,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Source {
+    path: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SourceBreakpoint {
+    pub line: usize,
+    pub column: Option<usize>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct SetBreakpointsArguments {
+    source: Source,
+    breakpoints: Option<Vec<SourceBreakpoint>>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Breakpoint {
+    pub id: Option<usize>,
+    pub verified: bool,
+    pub message: Option<String>,
+    pub source: Option<Source>,
+    pub line: Option<usize>,
+    pub column: Option<usize>,
+    pub end_line: Option<usize>,
+    pub end_column: Option<usize>,
+    pub instruction_reference: Option<String>,
+    pub offset: Option<usize>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct SetBreakpointsResponseBody {
+    breakpoints: Option<Vec<Breakpoint>>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct ContinueArguments {
+    thread_id: usize,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct ContinueResponseBody {
+    all_threads_continued: Option<bool>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct StackFrameFormat {
+    parameters: Option<bool>,
+    parameter_types: Option<bool>,
+    parameter_names: Option<bool>,
+    parameter_values: Option<bool>,
+    line: Option<bool>,
+    module: Option<bool>,
+    include_all: Option<bool>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct StackTraceArguments {
+    thread_id: usize,
+    start_frame: Option<usize>,
+    levels: Option<usize>,
+    format: Option<StackFrameFormat>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StackFrame {
+    id: usize,
+    name: String,
+    source: Option<Source>,
+    line: usize,
+    column: usize,
+    end_line: Option<usize>,
+    end_column: Option<usize>,
+    can_restart: Option<bool>,
+    instruction_pointer_reference: Option<String>,
+    // module_id
+    presentation_hint: Option<String>,
+}
```

Comment:
You'll want to define these types in a similar way to lsp-types: `<something>Params` `<something>Response` then a struct that implements `Request`:

```rust
#[derive(Debug)]
pub enum ApplyWorkspaceEdit {}

impl Request for ApplyWorkspaceEdit {
    type Params = ApplyWorkspaceEditParams;
    type Result = ApplyWorkspaceEditResponse;
    const METHOD: &'static str = "workspace/applyEdit";
}
```

Reply:
dap-types seems to have gone for a similar format: https://github.com/silvanshade/dap-types/blob/main/src/requests/client/run_in_terminal.rs#L52-L57

Their types look good but all the struct fields are private..

Reply:
Probably we can implement a necessary subset of DAP ourselves, keeping `helix-dap` a small low-dependency crate.

Reply:
Yeah I agree, we should still add a `Request` trait though. It makes it easy to validate that the right params & response are used since we can parameterize on `R::Params`, `R::Response`: https://github.com/helix-editor/helix/blob/9a39a10ddd5d8cf87e4ecedd54c4f4e06735ba33/helix-lsp/src/client.rs#L101-L111